extensions/vscode/src/eventErrors.ts

```diff
@@ -0,0 +1,46 @@
+// Copyright (C) 2024 by Posit Software, PBC.
+
+import { EventStreamMessage } from "./api/types/events";
+import { ErrorCode } from "./utils/errorTypes";
+
+export interface EventStreamMessageErrorCoded<T = Record<string, string>>
+  extends EventStreamMessage<T> {
+  errCode: ErrorCode;
+}
+
+export function isCodedEventErrorMessage(
+  msg: EventStreamMessage,
+): msg is EventStreamMessageErrorCoded {
+  return msg.errCode !== undefined;
+}
+
+type deploymentEvtErr = {
+  dashboardUrl: string;
+  localId: string;
+  error: string;
+};
+
+export const isEvtErrTargetNotFound = (
+  emsg: EventStreamMessageErrorCoded,
+): emsg is EventStreamMessageErrorCoded<deploymentEvtErr> => {
+  return emsg.errCode === "deploymentTargetNotFound";
+};
+
+export const isEvtErrTargetForbidden = (
+  emsg: EventStreamMessageErrorCoded,
+): emsg is EventStreamMessageErrorCoded<deploymentEvtErr> => {
+  return emsg.errCode === "deploymentTargetIsForbidden";
+};
+
+export const handleEventCodedError = (
+  emsg: EventStreamMessageErrorCoded,
+): string => {
+  if (isEvtErrTargetNotFound(emsg)) {
+    return `Content at ${emsg.data.dashboardUrl} could not be found. Please, verify the content "id" is accurate.`;
+  }
+  if (isEvtErrTargetForbidden(emsg)) {
+    return `You don't have enough permissions to deploy to ${emsg.data.dashboardUrl}. Please, verify the credentials in use.`;
+  }
+  const unknownErrMsg = emsg.data.error || emsg.data.message;
+  return `Unknown error: ${unknownErrMsg}`;
+};
```

Comment:
After a thorough review of this PR I think I understand @jonkeane's point a bit more about the codes for event messages.

These errors don't look to be doing the same amount of construction we expect the frontend to do for other errors (components, links, buttons, etc).

A highlight of this is the locked error. If you lock a piece of content and attempt to deploy you get a pretty good error even though it is unknown:

![CleanShot 2024-10-02 at 13 17 01@2x](https://github.com/user-attachments/assets/e13da7db-9b98-42b0-8f5c-84cdedf834af)

That didn't need a error code, if statements, or really much at all for it to make sense to the user.

Of course we could add a nice link to the content in the notification to adjust the lock setting, and if we want that we need a code to understand which error was passed up.

Reply:
That was the initial implementation a few weeks ago, the agent sending the error message and the extension only showing it. The argument of having the flexibility of sending only the error code so the UI can do many things is the approach we have now. Some error codes will require some construction on the UI while others will be fairly minimal.

IMO, at this point, consistency should be the key on where do we build this messages to avoid guessing games for each error message.

In this locked content example, I think it is a nice catch, it should have it's error code and be a known error and not show up as an `Unknown error` message.

Reply:
> IMO, at this point, consistency should be the key on where do we build this messages to avoid guessing games for each error message.

Completely agree. I lean toward consistency. I was trying to voice my understanding of some of the other concerns brought up, and think out loud how I'm reasoning about which to choose. It sounds like that didn't come across that way - sorry about that.

Reply:
No worries, glad to hear that. Sharing concerns and thinking out loud always helps on getting to common ground and understanding, I appreciate that 👍

Reply:
I will say that I expect we'll take more and more advantage of having the finely defined codes returned from the agent. I'm going to have to implement a richer error type coming back from the agent for deployment failure, so that I can catch when it has occurred due to TLS Certificate Verification failure, and offer the user the link to adjust the setting. Just wanted to add this context into the discussion. I also lean towards consistency here.